### ISSUE TITLE
Deep Learning

### ISSUE THREAD

**Issue by pillarliang** [pillarliang/blog#7]
# Introduction

机器学习把现实世界中的问题抽象成数学模型，通过求解数学模型进而解决现实生活中的问题。

机器学习研究和构建的是一种能够让计算机自己在数据中学习从而进行预测的特殊算法（**而非某一个特定的算法**）。

<img width="626" alt="11" src="https://github.com/pillarliang/blog/assets/28833267/3832b268-59d7-49e2-9289-d57f74cfdd79">

### 1. 设定范围-Find a set of function with unkown parameters

> 定出候选函数的集合（`**model**`）。参数不同，意味着函数不同
> 

深度学习中类神经网络的结构（例如：**CNN**，**RNN**，**Transformer**等）指的就是不同的候选**函数集合**。还有 Decision Tree 等。

模型的大小通常指模型中参数的数量，也可以指模型中神经元的数量。

数据量越小，设定的范围就越小。

> 当数据量非常小的时候，训练出来的模型容易过拟合，这意味着模型在训练数据上表现较好，但在测试数据上的表现很差。为了避免模型的过拟合，需要控制模型的复杂度，这时通常会采用更小的模型来训练。较简单的模型通常有更少的可调参数，因此可以更好地抵御过拟合。
<img width="777" alt="22" src="https://github.com/pillarliang/blog/assets/28833267/195b1ee2-10ea-41ed-8067-ff33d0f74646">

### 2. 设定目标-Define loss from training data

> 定出【评价函数好坏】的标准—— Loss 函数。
> 

Loss 函数的设定取决于训练资料。

- 如果是 supervised Learning，可以用预测的结果与真实值做计算。
- 如果是 Semi-supervised Learning，部分数据有label，部分数据没有。对于无label 的用数据，长得相似的一组数据它们的输出值应该接近，用相似数据输出值的差异来计算Loss.
- RL

[Loss and Error surface](https://www.notion.so/Loss-and-Error-surface-aa9bb98c6bc2475f900ae58d39366731?pvs=21)

### 3. 达成目标-optimization

> 在候选函数集（model） 中找出最好的函数

- [Gradient Descent](https://www.notion.so/Gradient-Descent-33c4e07694dd49b3a08ab0e6b919496a?pvs=21)
- [Backpropagation](https://www.notion.so/Backpropagation-2fc54e17c06d4bfa9c1e69914e84d09c?pvs=21)
- Genetic Algorithm
- …etc
- [Optimisation solution](https://www.notion.so/Optimisation-solution-b77961314ce34e1d8341ff51adc22059?pvs=21)

要注意当说到一个技术时，要能知道它是属于哪个范畴。

> ❌ RL 取代 DL。两个东西不是一个范畴。
> 

## 单层神经网络

[单层神经网络](https://www.notion.so/60cc8ddc232d4c20af9fd24d25d2eac2?pvs=21)

<img width="451" alt="44" src="https://github.com/pillarliang/blog/assets/28833267/02a40434-708b-4176-a4cd-c63d5b0f9448">

> 单层神经网络可能无法处理具有复杂非线性结构的数据，因此在深度学习中经常使用多层神经网络，如多层感知器（MLP）、卷积神经网络（CNN）和循环神经网络（RNN）等。

## 多层神经网络

多层神经网络就是先不把激活函数的输出 a 加在一起，而是再经过多层激活函数之后输出 a′，将 a′ 加在一起作为输出。每个激活函数（例如sigmoid、ReLU）都是一个neuron（神经元），每一隐藏层中包含多个神经元。所有的这些东西便组成了neural network（神经网络）。

隐藏层也不能太多，网络一旦太深，就容易产生 overfitting（过拟合）：在训练资料上有很好的表现，但是在测试资料上却表现很差，也就是在实际预测中没卵用！

<img width="455" alt="55" src="https://github.com/pillarliang/blog/assets/28833267/aec3d391-bea7-472b-b484-4d2feada28a1">

在一个全连接层（fully connected layer）或称为密集层（dense layer）中，每个神经元会接收前一层所有神经元的输出作为其输入（输入是一个向量，维度与前一层的神经元个数相同），然后进行加权求和，然后通过一个激活函数来产生一个数值输出。

## 优化措施

在优化中会遇到下面的问题：
<img width="815" alt="33" src="https://github.com/pillarliang/blog/assets/28833267/a9f8fa31-b72a-4139-8aa3-a63387a6d2f2">

### 1. Model bias
**what**: The model is too simple.

**solution**: redesign your model to make it more flexible.
- use more features
- use more neuros, layers

### 2. Optimisation issue

**what**: The model is good, but can’t find a parameter to make loss lower(梯度降不下去). ——> critical point

**solution:** 

[Optimisation solution](https://www.notion.so/Optimisation-solution-b77961314ce34e1d8341ff51adc22059?pvs=21)

**Distinguish between `model bias` and `optimization`:**

On the training data, increasing the layer of neuronal network can’t decrease the value of loss. This case is optimisation issue.

### 3. **Overfitting**

**what:** The performance on the *training data* is very good, but the results on the *testing data* are a mess and the loss is great.

**reason and solution**: 

[Overfitting & Underfitting](https://www.notion.so/Overfitting-Underfitting-82b87e00da1e4a1993cb7b7984851c8e?pvs=21)

**Trade off model bias and overfitting**:
Split your training data into training set and validation set for model selection.

### 4. **Mismatch**

**what:** Training data and testing data have different distribution.
